frontend/src/utils/SortUtils.ts

```diff
@@ -1,27 +1,89 @@
+import { useState, useMemo } from 'react';
+
 export enum SortingOrders {
   ASCENDING = 1,
   DESCENDING = -1,
 }
 
-export type SortKey<T, K> = (value: T) => K;
-export type SortComparison<T> = (a: T, b: T) => number;
+// from: https://stackoverflow.com/a/52991061
+type RequiredKeys<T> = {
+  [K in keyof T]-?: {} extends Pick<T, K> ? never : K;
+}[keyof T];
+
+// The type doesn't play nice with optional properties,
+// so those are just not allowed.
+// If optional property is needed for sorting, the key must be a function
+type PropertyKey<T, K> = {
+  [P in keyof T]: T[P] extends K ? P : never;
+}[RequiredKeys<T>];
+
+type Key<T, K> = PropertyKey<T, K> | ((value: T) => K);
+
+type Comparison<T> = (a: T, b: T) => number;
+
+export type Sort<T, K = any> =
```

Comment:
Would the name 'SortBy' suit this type better? Currently there's a sort function and Sort type, which may be a bit confusing

```suggestion
export type SortBy<T, K = any> =
```